jfix-stdlib-concurrency/src/main/java/ru/fix/stdlib/concurrency/futures/PendingFutureLimiter.java

```diff
@@ -0,0 +1,246 @@
+        }
+
+        this.thresholdFactor = thresholdFactor;
+
+        synchronized (counter) {
+            counter.notifyAll();
+        }
+
+        return this;
+    }
+
+    /**
+     * Неограниченное добавление future в очередь
+     */
+    public <T> CompletableFuture<T> enqueueUnlimited(CompletableFuture<T> future) throws InterruptedException {
+        return internalEnqueue(future, false);
+    }
+
+    /**
+     * Ограниченное добавление future в очередь.
+     * Если очередь уже достигла maxPendingCount то блокирует вызвавший поток до освобождения очереди,
+     * Далее, если maxFutureExecuteTime > 0, то по достижении этого времени выполнения в милисекундах,
+     *      future удаляется из очереди и создаётся возможность для добавления в неё новых future
+     * Если maxFutureExecuteTime = 0, до очередь не освобождается по таймауту и поставляющий в
+     * очередь поток будет заблокирован, пока future в очереди не выполнятся
+     */
+    public <T> CompletableFuture<T> enqueueBlocking(CompletableFuture<T> future) throws InterruptedException {
+        return internalEnqueue(future, true);
+    }
+
+    /**
+     * Регистрирует future во внутреннем счетчике.
+     * Подписывается на future listener который декрементит счетчик в случае завершения future.
+     * блокируется - если текущее значение внутреннего счетчика достигает или превышает maxPendingCount
+     * разблокируется - если количество pending futures = maxPendingCount * thresholdFactor
+     * разблокирование происходит в методе запущеном в ForkJoinPool после того как future будет завершена
+     * При наличии thresholdListener вызывает его методы в следующих случаях:
+     * ThresholdListener::onLimitReached - перед блокировкой
+     * ThresholdListener::onLimitSubceed - перед разблокировкой
+     */
+    protected <T> CompletableFuture<T> internalEnqueue(CompletableFuture<T> future,
+                                                       boolean needToBlock) throws InterruptedException {
+        if (counter.get() == maxPendingCount && thresholdListener != null) {
+            thresholdListener.onLimitReached();
+        }
+
+        if (needToBlock) {
+            awaitAndPurge();
+        }
+
+        counter.incrementAndGet();
+        pendingCompletableFutures.put(future, System.currentTimeMillis());
+        future.handleAsync((any, exc) -> {
+            if (exc != null) {
+                log.error(exc.getMessage(), exc);
+            }
+            long value = counter.decrementAndGet();
+            pendingCompletableFutures.remove(future);
+
+            if (value == 0 || value == getThreshold()) {
+                if (thresholdListener != null) {
+                    thresholdListener.onLimitSubceed();
+                }
+                synchronized (counter) {
+                    counter.notifyAll();
+                }
+            }
+            return null;
+        });
+        return future;
+    }
+
```

Comment:
There is a not  straight forward behavior in   awaitAndPurge, releaseAll, releaseEnqueued in terms of maxFutureExecuteTime.  
releaseAll, releaseEnqueued know about maxFutureExecuteTime but don't know that it can be 0. 
awaitAndPurge knows about maxFutureExecuteTime and that it can be 0 and also knows that releaseEnqueued do not check if maxFutureExecuteTime is 0.  

suggestion:

releaseTimeoutedFutures() - if maxFutureExecuteTime is 0 do not release anything and immediately returns, otherwise release timeouted futures. 

releaseAll rename to waitAll() - waits for all futures to complete normally or by timeout

awaitAndPurge rename to awaitOpportunityToEnqueueAndPurge

Reply:
> releaseAll rename to waitAll() - waits for all futures to complete normally or by timeout

Actual behavior of  this this method is not to wait, but just to release all queued futures. If we'll change the releaseEnqueued to releaseTimeoutedFutures and will exit if maxFutureExecuteTime is 0 it still will not wait for anything. 
I could change it to wait for all futures to complete, but i'm not sure if it makes sense, and how we'll do that? Just block the current thread until all the futures will be completed? 
Suggestion about this method - rename it to the releaseAllTimeoutedIfPossible().

About the rest comment - totally agree.

Reply:
Final decision - we need a method(s), which will wait all futures to complete successfully or to wait till timeout and then release. Thus, discussed method (reaseAll()) will be replaced by 
waitAll();//block invoked thread and wait all futures to complete
waitAll(long timeoutMillis); //block invoked thread and wait all futures to complete or till timeout is reached.